### ISSUE TITLE
Refactor num_parallel tracking in scheduler 

### ISSUE THREAD

**Issue by dhiltgen** [ollama/ollama#6078]
### What is the issue?

The current code passes a pointer and mutates it as we try to determine what the optimal parallel setting is, which makes the code hard to follow.

**Comment by Icelain** [comment]
Possibly a more readable approach would be to instead pass in numparallel to `pickBestFullFitByLibrary` and `pickBestPartialFitByLibrary` and have them return it back.
For example:
```g, numParallel := pickBestFullFitByLibrary(pending, ggml, gpus, numParallel)```
,
```gpus, numParallel = pickBestPartialFitByLibrary(pending, ggml, gpus, numParallel)```
